File: YukariLauncher/src/main/java/net/kdt/pojavlaunch/Logger.java
package net.kdt.pojavlaunch;

import androidx.annotation.Keep;

/**
 * 1つのファイルにログを記録するためのシングルトンクラス。
 * シングルトン部分は削除可能ですが、エンドデベロッパーによる実装が必要になります。
 */
@Keep
public class Logger {
    /**
     * 検閲されていない場合、テキストをログファイルに出力します。
     */
    public static native void appendToLog(String text);

    /**
     * ログファイルをリセットし、以前のログを消去します。
     */
    public static native void begin(String logFilePath);

    /**
     * ログをリッスンするための小さなリスナーインターフェース。
     */
    public interface eventLogListener {
        void onEventLogged(String text);
    }

    /**
     * ログリスナーをロガーにリンクします。
     */
    public static native void setLogListener(eventLogListener logListener);
}

File: YukariLauncher/src/main/java/com/arata/yukarilauncher/feature/log/Logger.kt
package com.arata.yukarilauncher.feature.log

import androidx.annotation.Keep

/**
 * ネイティブログ出力機能を提供するシングルトンオブジェクト。
 * JNIを介してC++側のログシステムと連携し、ログファイルへの書き込みや
 * ログイベントのリスナー通知を行います。
 */
@Keep
object Logger {
    @JvmStatic var enabled = true

    @JvmStatic external fun appendToLog(text: String)
    @JvmStatic external fun begin(logFilePath: String)
    @JvmStatic external fun setLogListener(logListener: eventLogListener?)

    fun interface eventLogListener {
        fun onEventLogged(text: String)
    }
}